demo-circuit/src/sc2sc/tests.rs

```diff
@@ -0,0 +1,565 @@
+use std::convert::TryInto;
+
+use algebra::Field;
+use cctp_primitives::{
+    commitment_tree::CommitmentTree,
+    type_mapping::{CoboundaryMarlin, FieldElement},
+};
+
+use primitives::FieldBasedMerkleTreePath;
+use r1cs_core::debug_circuit;
+use rand::{rngs::ThreadRng, thread_rng, Rng, RngCore};
+use rstest::*;
+use serial_test::serial;
+
+use crate::{
+    test_utils::{
+        self, init_g1_committer_key, CommitmentScBuilder, RandomWithdrawalCertificateDataBuilder,
+    },
+    GingerMHTBinaryPath,
+};
+
+use super::{Sc2Sc, ScCommitmentCertPath, MIN_CUSTOM_FIELDS, MSG_MT_HEIGHT};
+
+impl ScCommitmentCertPath {
+    /// Add implementation to simplify the extraction from a commitment
+    pub(crate) fn from_commitment(
+        cmt: &mut CommitmentTree,
+        sc_id: FieldElement,
+        cert_hash: FieldElement,
+    ) -> Option<Self> {
+        let cert_index = cmt
+            .get_cert_leaves(&sc_id)
+            .and_then(|certs| certs.iter().position(|h| h == &cert_hash))?;
+        Self::new(
+            cmt.get_cert_commitment(&sc_id)?,
+            cmt.get_fwt_commitment(&sc_id)?,
+            cmt.get_bwtr_commitment(&sc_id)?,
+            cmt.get_scc(&sc_id)?,
+            cmt.get_cert_merkle_path(&sc_id, cert_index)
+                .map(|p| p.try_into().expect("Should be a binary tree"))?,
+            cmt.get_sc_commitment_merkle_path(&sc_id)
+                .map(|p| p.try_into().expect("Should be a binary tree"))?,
+        )
+        .into()
+    }
+}
+
+fn assert_circuit(circuit: Sc2Sc, zk_rng: Option<&mut dyn RngCore>) {
+    let ck_g1 = init_g1_committer_key();
+    let setup_circuit = Sc2Sc::get_instance_for_setup(MIN_CUSTOM_FIELDS as u32);
+    let (pk, vk) = CoboundaryMarlin::index(&ck_g1, setup_circuit).unwrap();
+
+    assert_eq!(None, debug_circuit(circuit.clone()).unwrap());
+    let proof =
+        CoboundaryMarlin::prove(&pk, &ck_g1, circuit.clone(), zk_rng.is_some(), zk_rng).unwrap();
+    let public_inputs = [
+        circuit.next_sc_tx_commitments_root,
+        circuit.curr_sc_tx_commitments_root,
+        circuit.msg_hash,
+    ];
+    assert!(CoboundaryMarlin::verify(&vk, &ck_g1, &public_inputs, &proof).unwrap());
+}
+
+#[fixture]
+fn rng() -> ThreadRng {
```

Comment:
Same comment related to use of ThreadRng instead of a generic R: Rng

Reply:
The last one... :fireworks: